Avoid defining RpcCUDAFuture subclass in TensorPipe agent

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56518 Add tests for CUDAFuture
* #56517 Add CUDA support to a user-created torch.futures.Future
* #56516 Support custom Python classes in CUDAFuture
* #56515 Allow to specify a set of device for CUDAFuture
* #56514 Split CUDAFuture into a .h and a .cpp file
* **#56513 Avoid defining RpcCUDAFuture subclass in TensorPipe agent**
* #56512 Merge ivalue::Future's markCompleted and markCompletedWithDataPtrs
* #56511 Gracefully handle failure of DataPtr extraction in CUDAFuture
* #56510 Raise TypeErrors when IValue::getSubValues fails

The RpcCUDAFuture class existed solely to support extracting DataPtrs from a Message class. However, this can be done more simply by using a vanilla CUDAFuture and just extracting those DataPtrs before marking it complete and passing them to markCompleted.

This allows to make the DataPtr extraction logic of CUDAFuture private again.

Differential Revision: [D27861064](https://our.internmc.facebook.com/intern/diff/D27861064/)

lw committed:
Avoid defining RpcCUDAFuture subclass in TensorPipe agent

The RpcCUDAFuture class existed solely to support extracting DataPtrs from a Message class. However, this can be done more simply by using a vanilla CUDAFuture and just extracting those DataPtrs before marking it complete and passing them to markCompleted.

This allows to make the DataPtr extraction logic of CUDAFuture private again.

Differential Revision: [D27861064](https://our.internmc.facebook.com/intern/diff/D27861064/)

[ghstack-poisoned]

mrshenli reviewed (approved):
Thanks for cleaning this up!

lw committed:
Update on "Avoid defining RpcCUDAFuture subclass in TensorPipe agent"


The RpcCUDAFuture class existed solely to support extracting DataPtrs from a Message class. However, this can be done more simply by using a vanilla CUDAFuture and just extracting those DataPtrs before marking it complete and passing them to markCompleted.

This allows to make the DataPtr extraction logic of CUDAFuture private again.

Differential Revision: [D27861064](https://our.internmc.facebook.com/intern/diff/D27861064/)

[ghstack-poisoned]
